envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/ClientNameTrustedHeaderTest.kt

```diff
@@ -0,0 +1,171 @@
+        )
+
+        @JvmField
+        @RegisterExtension
+        val envoy5 = EnvoyExtension(envoyControl, service, Echo5EnvoyAuthConfig.copy(configOverride = echoClientsConfig))
+    }
+
+    @BeforeEach
+    fun beforeEach() {
+        consul.server.operations.registerService(
+            id = "echo",
+            name = "echo",
+            address = envoy.container.ipAddress(),
+            port = EnvoyContainer.INGRESS_LISTENER_CONTAINER_PORT,
+            tags = listOf("mtls:enabled")
+        )
+        waitForEnvoysInitialized()
+    }
+
+    private fun waitForEnvoysInitialized() {
+        untilAsserted {
+            assertThat(envoy2.container.admin().isEndpointHealthy("echo", envoy.container.ipAddress())).isTrue()
+        }
+    }
+
+    @Test
+    fun `should always remove "x-client-name-trusted" header on every envoy ingress request`() {
+        untilAsserted {
+            // when
+            val response = envoy2.ingressOperations.callLocalService(
+                "/log-unlisted-clients",
+                Headers.of(mapOf("x-client-name-trusted" to "fake-service"))
+            )
+            // then
+            assertThat(response.header("x-client-name-trusted")).isNull()
+        }
+    }
+
+    @Test
+    fun `should add "x-client-name-trusted" header on envoy ingress request`() {
+        untilAsserted {
+            // when
+            val response = envoy2.egressOperations.callService("echo", emptyMap(), "/log-unlisted-clients")
+            // then
+            assertThat(response.header("x-client-name-trusted")).isEqualTo("echo2")
+        }
+    }
+
+    @Test
+    fun `should override "x-client-name-trusted" header with trusted client name form certificate on request`() {
+        untilAsserted {
+            // when
+            val headers = mapOf("x-client-name-trusted" to "fake-service")
+            val response = envoy2.egressOperations.callService("echo", headers, "/log-unlisted-clients")
+            // then
+            assertThat(response.header("x-client-name-trusted")).isEqualTo("echo2")
+        }
+    }
+
+    @Test
+    fun `should set "x-client-name-trusted" header based on all URIs in certificate SAN field`() {
+        untilAsserted {
+            // when
+            val response = envoy4.egressOperations.callService("echo", emptyMap(), "/log-unlisted-clients")
+            // then
+            assertThat(response.header("x-client-name-trusted")).isEqualTo("echo4,echo4-special,echo4-admin")
+        }
+    }
+
+    @Test
+    fun `should set "x-client-name-trusted" header based on URIs in certificate SAN field regardles protocol used in SAN alt name`() {
+        untilAsserted {
+            // when
+            val response = envoy5.egressOperations.callService("echo", emptyMap(), "/log-unlisted-clients")
+            // then
+            assertThat(response.header("x-client-name-trusted")).isEqualTo("echo5,echo5-special,echo5-admin")
+        }
+    }
```

Comment:
We need tests for:
* request with valid SAN URI but not validated certificate (for example: signed by other CA)
* request with slightly invalid SAN URI and certificate validated correctly

Reply:
ad 1. It's not needed, we already have tests for it. It turns out, we already have a protection against clients with invalid certificates. Even if `require-client-certificate` is set to `false`, when a client presents a certificate, it is validated and rejected if invalid. So we don't need additional check in lua. The check in lua (`peerCertificateValidated()`) is bugged - it sometimes returns false even if certificate is valid.

ad 2. done